swaybar: support gestures bindings

FlorianFranzen opened the pull request:
**The first half of this PR was already merged, but the swaybar implementation has to be reworked to not just implement what the base implementation in sway already provides, i.e. we probably want more usable per-block gestures.**

This is a continuation of the work of @michaelweiser  in #4952. It extends the previous work by adding a `bindgesture` and `unbindgesture` commands to sway and swaybar and extending support from `swipe` to `hold` and `pinch` gestures.

Gestures are specified by the type and optionally by the number of fingers and the directions, i.e. as `<type>[:fingers][:directions]`.

Directions can only be determined at the and of the gesture and for `pinch` types can be a combination of multiple gestures. Available gestures are `up`, `down`, `left` and `right`, as well as `inward`, `outward`, `clockwise` and `counterclockwise`.

Any gesture that does not match any bindings will be forwarded to the client.

A good overview of the data available from gestures can be found [here](https://wayland.freedesktop.org/libinput/doc/latest/gestures.html).

FlorianFranzen committed:
sway/bar: add bindgesture support

FlorianFranzen commented:
Rebased on latest master, removes conflict with #6775.

rpigott reviewed (changes requested):
First pass review, haven't gotten to the swaybar stuff yet. I'll take a look at the rest soon enough, but it's looking sensible so far. I like the syntax of the new command as well.

I've waffled enough in the past about whether or not compositor captured gestures belong in a separate non-default seatop, but as long as I can't figure out how to crash it I think it will be fine.

Thanks for yours and @michaelweiser's work on this.

David96 commented:
@FlorianFranzen Are you planning on continuing this pull request? I quite like the ability to bind gestures and if not I might be able to pick it up (even if probably at a slow pace :sweat_smile:)
For starters, I rebased it onto the newest master (https://github.com/David96/sway/commits/pointer-gestures)

FlorianFranzen commented on common/gesture.c:
```diff
@@ -0,0 +1,348 @@
+
+const uint8_t GESTURE_FINGERS_ANY = 0;
+
+/// Helper to easily allocate and format string
```
Sorry, habit from using doxygen.

RPigott replied:
The comments are good, but as we don't use doxygen I'd drop the special formatting.

FlorianFranzen commented on include/stringop.h:
```diff
@@ -2,6 +2,7 @@
 #define _SWAY_STRINGOP_H
 
 #include <stdbool.h>
+#include <stddef.h>
```
No, the stringop header uses definitions from `stddef.h` and can not be included, unless `stddef.h` is included first.

RPigott replied:
Oh, ok. Good catch.

FlorianFranzen commented on common/gesture.c:
```diff
@@ -0,0 +1,348 @@
+
+/// Helper to easily allocate and format string
+static char *strformat(const char *format, ...) {
+	char *result = malloc(256);
```
Ah, of course. Thanks for the tip!

RPigott commented on sway/input/seatop_default.c:
```diff
@@ -753,215 +752,322 @@ static void handle_pointer_axis(struct sway_seat *seat,
+}
+
+/// Small wrapper around seat_execute_command to work on gesture bindings
+static void gesture_binding_execute(struct sway_seat *seat,
```
Just reading my old comments here, and I think such a flag should not be included. If there is a desire for it we can add it later.

FlorianFranzen replied:
Ok, I will have another look, just so I fully understand this. And we receive a surface when a gesture begins, so it does make sense to use that information. I won't spent to much time on this.

FlorianFranzen replied:
So it seems the main difference is that mouse bindings provide `execute_command` with an additional container pointer in `seat_execute_command`. It seems however the context of this container is only relevant, if you do not provide any criteria, so I am not sure your example applies, or ever stopped working, because criteria are handled in `execute command` either way. So I added that same behavior for gesture events. I did not use the surface passed to at the beginning of a gesture, and I think that is more an issue for another PR.

FlorianFranzen commented:
@RPigott Rebased onto master and addressed any review comments that could be handled quickly. Will have another look at the matching, overwriting and pointer context again.

@David96 Thank you for your effort. I ended fixing a few other things during my rebase.

FlorianFranzen commented on sway/commands/gesture.c:
```diff
@@ -0,0 +1,166 @@
+/**
+ * Add gesture binding to config
+ */
+static struct cmd_results *gesture_binding_add(
```
If not overwritten, it is added to the list just below in the `if (!overwritten)` conditional.

FlorianFranzen commented:
@RPigott I addressed the remaining issues, would love to see this merged.

@emersion I know you reviewed the last gesture support PR, would be great if you could take a quick look at this one as well.

RPigott commented on sway/input/seatop_default.c:
```diff
@@ -866,7 +866,7 @@ static struct sway_gesture_binding* gesture_tracker_end_and_match(
 // Small wrapper around seat_execute_command to work on gesture bindings
 static void gesture_binding_execute(struct sway_seat *seat,
-									struct sway_gesture_binding *binding) {
+			struct sway_gesture_binding *binding) {
```
Just two tabs will do here, the style is described in the CONTRIBUTING doc.

rpigott reviewed (changes requested):
Okay I've reviewed the swaybar stuff now too, I think the logic is alright here just style comments.

I'll try this out and assuming nothing comes up once those comments are addressed I can approve.

FlorianFranzen commented on common/gesture.c:
```diff
@@ -0,0 +1,352 @@
+struct gesture *gesture_tracker_end(struct gesture_tracker *tracker) {
+	struct gesture *result = calloc(1, sizeof(struct gesture));
+
+	// Ignore gesture under some threshold
+	// TODO: Make configurable
+	const double min_rotation = 5;
+	const double min_scale_delta = 0.1;
```
@RPigott What do you think about these? I probably should make them configurable.

emersion replied:
New configuration entries will need to be maintained/supported forever because Sway's config file format is stable. I'd prefer hardcoding values for now, try to find better values if this is an issue, and only add a config option if it turns out we have no choice.

FlorianFranzen replied:
I guess requiring at least a five degree rotation and at least a 10% scale change is reasonable.

FlorianFranzen commented on sway/commands/gesture.c:
```diff
@@ -0,0 +1,166 @@
+/**
+ * Returns true if the bindings have the same gesture type, direction, etc
+ */
+static bool binding_gesture_equal(struct sway_gesture_binding *binding_a,
```
Input and flags are checked here. And unbingesture uses this function to find the gesture to unbind. So it should all be accounted for.

FlorianFranzen commented:
Seems `scratchpad show` generally does not work well with regular mouse binding, because it silently does nothing, if the provided source hint is not in the scratchpad. I worked around it by adding a `[floating]` which works, but certainly is weird. Maybe I should add another flag after all.

RPigott commented:
> However, since I started treating gesture binding more like mouse bindings during execution (https://github.com/swaywm/sway/commit/36809ec53da5106707dc8d02fc52266e66b78908), scratchpad show now only works on a workspace surface and no longer just anywhere, so I had to add the [floating]to make it work anywhere again.

Sorry, I should have reviewed that commit as well. When you added a context node, the executor in commands.c considers it to be a criteria overridden node and only attempts to show the context node from the scratchpad. If the cursor is over a container, it is never a scratchpad hidden container, so the command won't function as you expect.

When I said we don't need an option, I meant that it was fine to keep the keyboard focused behavior you already had, I just wasn't sure if the behavior change was intentional. I don't think there is a good answer to what the default context node should be for gesture bindings so its totally fine to keep it simple until someone complains it doesn't behave as expected. @FlorianFranzen  just drop that commit and it should be fine.

trbjo commented:
FWIW I have also been testing this PR recently. It works really well.
You can use these bindgestures to achieve three finger forward/backward navigation:
```bindgesture swipe:3:right seat seat0 cursor press button8, seat seat0 cursor release button8
bindgesture swipe:3:left seat seat0 cursor press button9, seat seat0 cursor release button9
```

RPigott commented:
One last issue to sort out is the rebase. I think ideally we shouldn't introduce the key-enum based syntax only to immediately remove it in favor of the bindgesture command, so it would be best to rebase these changes to eliminate the old key-enums from the history. But in this case it's a little complicated because I'd also like to preserve @michaelweiser's authorship where possible.

@FlorianFranzen it looks to me like it might be difficult to split the changes into commits by author, so I think it would be acceptable just squash as necessary to remove the old syntax and include a "Co-authored-by:" commit tag (https://git.wiki.kernel.org/index.php/CommitMessageConventions) where appropriate.

michaelweiser commented:
I'm not married to that authorship at all. Feel free to merge commits as makes most sense logically.

FlorianFranzen committed:
swaybar: add bindgesture support

Co-authored-by: Michael Weiser <michael.weiser@gmx.de>

FlorianFranzen commented:
Sorry about that, forgot to reset authors in the squash. Fixed now.

rpigott reviewed (approved):
This LGTM! Thanks for your hard work!

emersion commented:
Thanks a lot for reviewing this @RPigott! I'll have a quick look at this next week. Please ping me if I forget.

emersion commented on sway/commands/bar/gesture.c:
```diff
@@ -0,0 +1,111 @@
+	for (int i = 0; i < gestures->length; ++i) {
+		struct bar_gesture *current = gestures->items[i];
+		if (gesture_equal(&current->gesture, &binding->gesture)) {
+
```
Style nit: unnecessary newline

emersion reviewed (changes requested):
Here is one last round of comments, mostly style fixes. I haven't looked at the logic very closely but I trust @RPigott's review and on a high-level that all looks good!

Can the `swaybar-protocol` man page be updated as well?

FlorianFranzen commented:
> I've addressed all of these comments, except the IPC man page update. I'm having second thoughts about including gestures in `GET_BAR_CONFIG`, it's inconsistent with click events and it will make it impossible to implement features such as per-block gestures (e.g. swipe on music block to skip current track).

How about just merge gesture support for sway itself then decide how to handle the bar in a separate PR? I would like to see this merged sooner then later.

FlorianFranzen commented:
Great, thank you. I rebased the branch and added a disclaimer.

andraxin commented:
@FlorianFranzen 
I've been using the original gesture patches for what it feels like time immemorial.

When switching to Sway 1.8, I noticed the "official" gesture support with a dedicated
keyword `bindgesture` instead of `bindsym`.  Since this causes loss of functionality
compared to the original patch (modifiers are no longer supported), I'm wondering
* if this change was necessary and why
  - personally, I don't particularly care what it's called, but I have several actions
    configured *with* modifiers that no longer work
* if adding modifier support to `bindgesture` would be difficult

BTW, don't get me wrong, I'm grateful for the work you (and @michaelweiser) put in
to get this _crucial_ functionality accepted; I'm just hoping to get my modifier+gesture
actions back soonish... :smirk:

andraxin commented:
@FlorianFranzen

---
I'm not sure _where_ to put this comment...
The original PR (https://github.com/swaywm/sway/pull/4952) is closed.
This one is partially merged, and based on its title has a different focus.
Unable to resolve my confusion, I added this same comment in **both**.
Please excuse the duplication! 

---

I've been using the original gesture patches for what it feels like time immemorial.

When switching to Sway 1.8, I noticed the "official" gesture support with the dedicated
keyword `bindgesture` instead of bindsym. Since this causes loss of functionality
compared to the original patch (modifiers are no longer supported), I'm wondering

*    ~~if this change was necessary and why~~
     -   scratch that; I found @rpigott 's comment [above](https://github.com/swaywm/sway/pull/4952#issuecomment-940810475) 
     -   personally, I don't particularly care what it's called, but I have several actions
         configured with modifiers that no longer work
 *   if adding modifier support to bindgesture would be difficult
     - a quick look in the code tells me that this isn't entirely trivial;
       reverting to Sway 1.7 + patches for now is much easier. :smiling_face_with_tear: 

BTW, don't get me wrong, I'm grateful for the work you (and @michaelweiser) put in
to get this crucial functionality accepted; I'm just hoping to get my modifier+gesture
actions back soonish...  :smirk:

Another minor annoyance with `bindgesture`is that `mouse_warping` no longer seems
to work with gestures the way it did when they were specified with `bindsym`.
